otelconf/metric.go

```diff
@@ -0,0 +1,591 @@
+		}
+		return periodicExporter(ctx, r.Periodic.Exporter, opts...)
+	}
+
+	if r.Pull != nil {
+		return pullReader(ctx, r.Pull.Exporter)
+	}
+	return nil, newErrInvalid("no valid metric reader")
+}
+
+func pullReader(ctx context.Context, exporter PullMetricExporter) (sdkmetric.Reader, error) {
+	if exporter.PrometheusDevelopment != nil {
+		return prometheusReader(ctx, exporter.PrometheusDevelopment)
+	}
+	return nil, newErrInvalid("no valid metric exporter")
+}
+
+func periodicExporter(ctx context.Context, exporter PushMetricExporter, opts ...sdkmetric.PeriodicReaderOption) (sdkmetric.Reader, error) {
+	exportersConfigured := 0
+	var exportFunc func() (sdkmetric.Reader, error)
+
+	if exporter.Console != nil {
+		exportersConfigured++
+		enc := json.NewEncoder(os.Stdout)
+		enc.SetIndent("", "  ")
+
+		exp, err := stdoutmetric.New(
+			stdoutmetric.WithEncoder(enc),
+		)
+		if err != nil {
+			return nil, err
+		}
+		exportFunc = func() (sdkmetric.Reader, error) {
+			return sdkmetric.NewPeriodicReader(exp, opts...), nil
+		}
+	}
+	if exporter.OTLPHttp != nil {
+		exportersConfigured++
+		exp, err := otlpHTTPMetricExporter(ctx, exporter.OTLPHttp)
+		if err != nil {
+			return nil, err
+		}
+		exportFunc = func() (sdkmetric.Reader, error) {
+			return sdkmetric.NewPeriodicReader(exp, opts...), nil
+		}
+	}
+	if exporter.OTLPGrpc != nil {
+		exportersConfigured++
+		exp, err := otlpGRPCMetricExporter(ctx, exporter.OTLPGrpc)
+		if err != nil {
+			return nil, err
+		}
+		exportFunc = func() (sdkmetric.Reader, error) {
+			return sdkmetric.NewPeriodicReader(exp, opts...), nil
+		}
+	}
+	if exporter.OTLPFileDevelopment != nil {
+		// TODO: implement file exporter https://github.com/open-telemetry/opentelemetry-go/issues/5408
+		return nil, newErrInvalid("otlp_file/development")
+	}
+
+	if exportersConfigured > 1 {
+		return nil, newErrInvalid("must not specify multiple exporters")
+	}
+
+	if exportFunc != nil {
+		return exportFunc()
+	}
+
+	return nil, newErrInvalid("no valid metric exporter")
+}
+
+func otlpHTTPMetricExporter(ctx context.Context, otlpConfig *OTLPHttpMetricExporter) (sdkmetric.Exporter, error) {
+	opts := []otlpmetrichttp.Option{}
+
+	if otlpConfig.Endpoint != nil {
+		u, err := url.ParseRequestURI(*otlpConfig.Endpoint)
+		if err != nil {
+			return nil, errors.Join(newErrInvalid("endpoint parsing failed"), err)
+		}
+		opts = append(opts, otlpmetrichttp.WithEndpoint(u.Host))
+
+		if u.Scheme == "http" {
+			opts = append(opts, otlpmetrichttp.WithInsecure())
+		}
+		if u.Path != "" {
+			opts = append(opts, otlpmetrichttp.WithURLPath(u.Path))
+		}
+	}
+	if otlpConfig.Compression != nil {
+		switch *otlpConfig.Compression {
+		case compressionGzip:
+			opts = append(opts, otlpmetrichttp.WithCompression(otlpmetrichttp.GzipCompression))
+		case compressionNone:
+			opts = append(opts, otlpmetrichttp.WithCompression(otlpmetrichttp.NoCompression))
```

Comment:
Is NoCompression the default, or is that only for grpc?

Reply:
Right, i guess the default value for compression for http is also `NoCompression` based on the default value of the `Compression` enum. Probably good to be more specific than not in case that ever changes in the future.

Might be worth updating the grpc code to specifically set the compressor as well for consistency. Looking at the implementations for otlpgrpc for logs, metrics, and traces, the code does different things that end up setting the NoCompression option. Happy to follow this up in another PR if you're ok with it

Reply:
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#:~:text=If%20no%20compression%20value%20is%20explicitly%20specified%2C%20SIGs%20can%20default%20to%20the%20value%20they%20deem%20most%20useful%20among%20the%20supported%20options.%20This%20is%20especially%20important%20in%20the%20presence%20of%20technical%20constraints%2C%20e.g.%20directly%20sending%20telemetry%20data%20from%20mobile%20devices%20to%20backend%20servers.

> If no compression value is explicitly specified, SIGs can default to the value they deem most useful among the supported options. This is especially important in the presence of technical constraints, e.g. directly sending telemetry data from mobile devices to backend servers.